src/wallet/wallet.cpp

```diff
@@ -4021,14 +4057,57 @@ CWallet* CWallet::CreateWalletFromFile(const std::string& name, const fs::path&
         walletInstance->SetMaxVersion(nMaxVersion);
     }
 
+    // Upgrade to HD if explicit upgrade
+    if (gArgs.GetBoolArg("-upgradewallet", false))
+    {
+        LOCK(walletInstance->cs_wallet);
+
+        // Do not upgrade versions to any version between HD_SPLIT and FEATURE_PRE_SPLIT_KEYPOOL unless already supporting HD_SPLIT
+        int nMaxVersion = walletInstance->nWalletVersion;
+        if (!walletInstance->CanSupportFeature(FEATURE_HD_SPLIT) && nMaxVersion >=FEATURE_HD_SPLIT && nMaxVersion < FEATURE_PRE_SPLIT_KEYPOOL) {
+            InitError(_("Cannot upgrade a non HD split wallet without upgrading to support pre split keypool"));
```

Comment:
This error would scare me (since I'm juggling privkeys and it's not a very easy to understand message). Maybe be nice and do `[...]split keypool. To do this, run [XYZ] first.` or similar?

Reply:
I edited the error to say what to do instead.